File: Data Structures & Algorithms/counting-bits/submission-5.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> res(n+1,0);
        int offset = 1;
        for(int i=1;i<=n;i++) {
            if(offset*2 == i) offset = i;
            res[i] = 1 + res[i-offset];
        }
        return res;
    }
};

/*
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
|
0 1 1 2 1 2 2 3 1 2 2  3  2  3  3  4

8*2 == 16 -- 
dp[8] = 1 + dp[0]
*/

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-5.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        Stack = [(-1, 0)]
        res = 0
        for i, num in enumerate(heights):
            temp = i
            while Stack[-1][0] >= num:
                height, index = Stack.pop()
                res = max(res, height*(i-index))
                temp = index
            
            Stack.append((num, temp))
        
        n = len(heights)
        while len(Stack) > 1:
            height, index = Stack.pop()
            res = max(res, height*(n-index))
        
        return res
